learn/best_practices/type_casting.md

```diff
@@ -0,0 +1,494 @@
+
+end module
+```
+
+Usage:
+
+``` fortran
+subroutine foo(a, k)
+use integrals, only: simpson
+real(dp) :: a, k
+print *, simpson(f, 0._dp, pi)
+print *, simpson(f, 0._dp, 2*pi)
+
+contains
+
+real(dp) function f(x) result(y)
+real(dp), intent(in) :: x
+y = a*sin(k*x)
+end function f
+
+end subroutine foo
+```
+
+Using type(c\_ptr) Pointer
+--------------------------
+
+In C, one would use the `void *` pointer. In Fortran, one can use
+`type(c_ptr)` for exactly the same purpose.
+
+Integrator:
+
+``` fortran
+module integrals
+  use types, only: dp
+  use iso_c_binding, only: c_ptr
+  implicit none
+  private
+  public simpson
+
+contains
+
+real(dp) function simpson(f, a, b, data) result(s)
+  real(dp), intent(in) :: a, b
+  interface
+    real(dp) function func(x, data)
+    use types, only: dp
+    implicit none
+    real(dp), intent(in) :: x
+    type(c_ptr), intent(in) :: data
+    end function
+  end interface
+  procedure(func) :: f
+  type(c_ptr), intent(in) :: data
+  s = (b-a) / 6 * (f(a, data) + 4*f((a+b)/2, data) + f(b, data))
+end function
+
+end module
+```
+
+Usage:
+
+``` fortran
+module test
+  use types, only: dp
+  use integrals, only: simpson
+  use iso_c_binding, only: c_ptr, c_loc, c_f_pointer
+  implicit none
+  private
+  public foo
+
+  type f_data
+    ! Only contains data that we need for our particular callback.
+    real(dp) :: a, k
+  end type
+
+contains
+
+real(dp) function f(x, data) result(y)
+  real(dp), intent(in) :: x
+  type(c_ptr), intent(in) :: data
+  type(f_data), pointer :: d
+  call c_f_pointer(data, d)
+  y = d%a * sin(d%k * x)
+end function
+
+subroutine foo(a, k)
+  real(dp) :: a, k
+  type(f_data), target :: data
+  data%a = a
+  data%k = k
+  print *, simpson(f, 0._dp, pi, c_loc(data))
+  print *, simpson(f, 0._dp, 2*pi, c_loc(data))
+end subroutine
+
+end module
+```
+
+As always, with the advantages of such re-casting, as Fortran lets you
+do if you really want to, come also the disadvantages that fewer
+compile- and run-time checks are possible to catch errors; and with
+that, inevitably more leaky, bug-prone code. So one always has to
+balance the costs and benefits.
+
+Usually, in the context of scientific programming, where the main thrust
+is to represent and solve precise mathematical formulations (as opposed
+to create a GUI with some untold number of buttons, drop-downs, and
+other interface elements), simplest, least bug-prone, and fastest is to
+use one of the previous approaches.
+
+transfer() Intrinsic Function
+-----------------------------
+
+Before Fortran 2003, the only way to do type casting was using the
+`transfer` intrinsic function. It is functionally equivalent to the
+method V, but more verbose and more error prone. It is now obsolete and
+one should use the method V instead.
+
+Examples:
+
+<http://jblevins.org/log/transfer>
+
+<http://jblevins.org/research/generic-list.pdf>
+
+<http://www.macresearch.org/advanced_fortran_90_callbacks_with_the_transfer_function>
```

Comment:
This doesn't look good at all, either we describe it ourselves or we remove this section completely.